### PR TITLE
introduce cached_properties in GPU methods classes (no more lengthy constructors + bonuses from lazy evaluation + implies dropping support for Python 3.7) + some C code syntax fixes & cleanups + atomicAdd workaround for double type on older CUDA

### PR DESCRIPTION
**Pull request by slayoo**

### CODE REVIEW

**Comment by abulenok** [comment]
@slayoo Why do we drop python 3.7 support?

File: tests/unit_tests/backends/test_freezing_methods.py [hunk 1]
```diff
@@ -96,8 +96,12 @@ def test_freeze_singular(backend_class):
         )
 
     @staticmethod
+    @pytest.mark.parametrize("double_precision", (True, False))
     # pylint: disable=too-many-locals
-    def test_freeze_time_dependent(backend_class, plot=False):
+    def test_freeze_time_dependent(backend_class, double_precision, plot=False):
+        if backend_class.__name__ == "Numba" and not double_precision:
+            pytest.skip()
```
Comment (abulenok) [inline tests/unit_tests/backends/test_freezing_methods.py:103]: maybe let's create an issue for that?

**Review by abulenok — APPROVED** [review]
LGTM, I like the new approach with `@cached_property`, great idea!

File: tests/unit_tests/backends/test_freezing_methods.py (hunk 1 above)
Comment (slayoo) [reply tests/unit_tests/backends/test_freezing_methods.py:103]: good point! https://github.com/open-atmos/PySDM/issues/1144 (will be adding TODO labels as well)

**Comment by slayoo** [comment]
> @slayoo Why do we drop python 3.7 support?

@abulenok 
First, `@cached_property` was introduced in Python 3.8.
Second, Python 3.7 has reached "end of life" in June 2023: https://devguide.python.org/versions/

**Pull request merged**